js-api-spec/compiler.test.ts

```diff
@@ -0,0 +1,176 @@
+import {initAsyncCompiler, initCompiler, SassString} from 'sass';
+
+import {sandbox} from './sandbox';
+import {spy, URL} from './utils';
+
+const functions = {'foo($args)': (args: unknown) => new SassString(`${args}`)};
+const importers = [
+  {
+    canonicalize: (url: string) => new URL(`u:${url}`),
+    load: (url: typeof URL) => ({
+      contents: `.import {value: ${url.pathname}} @debug "imported";`,
+      syntax: 'scss' as const,
+    }),
+  },
+];
+const asyncImporters = [
+  {
+    canonicalize: (url: string) =>
+      Promise.resolve(importers[0].canonicalize(url)),
+    load: (url: typeof URL) => Promise.resolve(importers[0].load(url)),
+  },
+];
+const getLogger = () => ({debug: spy(() => {})});
+
+describe('Compiler', () => {
+  let compiler: Compiler;
+
+  beforeEach(() => {
+    compiler = initCompiler();
+  });
+
+  afterEach(() => {
+    compiler.dispose();
+  });
+
+  describe('compileString', () => {
+    it('performs complete compilations', () => {
+      const logger = getLogger();
+      const result = compiler.compileString(
+        '@import "bar"; .fn {value: foo(baz)}',
+        {importers, functions, logger}
+      );
+      expect(result.css).toEqualIgnoringWhitespace(
+        '.import {value: bar;} .fn {value: "baz";}'
+      );
+      expect(logger.debug).toHaveBeenCalledTimes(1);
+    });
+
+    it('throws after being disposed', () => {
+      compiler.dispose();
+      expect(() => compiler.compileString('$a: b; c {d: $a}')).toThrowError();
+    });
+  });
+
+  describe('compile', () => {
+    it('performs complete compilations', () =>
+      sandbox(dir => {
+        const logger = getLogger();
+        dir.write({'input.scss': '@import "bar"; .fn {value: foo(bar)}'});
+        const result = compiler.compile(dir('input.scss'), {
+          importers,
+          functions,
+          logger,
+        });
+        expect(result.css).toEqualIgnoringWhitespace(
+          '.import {value: bar;} .fn {value: "bar";}'
+        );
+        expect(logger.debug).toHaveBeenCalledTimes(1);
+      }));
+
+    it('throws after being disposed', () =>
+      sandbox(dir => {
+        dir.write({'input.scss': '$a: b; c {d: $a}'});
+        compiler.dispose();
+        expect(() => compiler.compile(dir('input.scss'))).toThrowError();
+      }));
+  });
+});
+
+describe('AsyncCompiler', () => {
+  let compiler: AsyncCompiler;
+
+  beforeEach(async () => {
+    compiler = await initAsyncCompiler();
+  });
+
+  afterEach(async () => {
+    await compiler.dispose();
+  });
+
+  describe('compileStringAsync', () => {
+    it('handles multiple concurrent compilations', async () => {
+      const logger = getLogger();
+      const compilations = Array(5)
+        .fill(0)
+        .map((_, i) =>
+          compiler.compileStringAsync(
+            `@import "${i}"; .fn {value: foo(${i})}`,
+            {importers: asyncImporters, functions, logger}
+          )
+        );
```

Comment:
> verify that the async compiler in particular gracefully handles multiple concurrent compilations, especially when those compilations are "chatty" (calling custom functions, importers, and loggers)

@nex3 this will fire up 5 "chatty" compilations. Let me know what you think

Reply:
I think we should probably run more like 1000 of them. We really want to pressure-test the concurrency here.

Reply:
Updated to 1000 in https://github.com/sass/sass-spec/pull/1954/commits/36db21b77999bb27cf7085cde9e7b88b8c5c0252. In local testing I was able to go to about 2000 before Jasmine's 5 second timeout kicked in

Reply:
It seems like all Mac OS runs (even on other branches) are taking around double the time to complete when compared to Ubuntu. @nex3 would you be open to increasing the timeout for this test only on macs?

Reply:
I increased the timeouts to 15 seconds for string tests and 40 seconds for node tests. See https://github.com/sass/sass-spec/pull/1954/commits/286e7fafd73d3428b8e51d85beb10a097d1bed18